juju/model.py

```diff
@@ -0,0 +1,166 @@
+        to self.relations[relation_name][0].
+        """
+        num_related = len(self.relations[relation_name])
+        if num_related > 1:
+            raise TooManyRelatedApps(relation_name, num_related, 1)
+        elif num_related == 0:
+            return None
+        else:
+            return self.relations[relation_name][0]
+
+
+class Application:
+    def __init__(self, name, units):
+        # TODO Do we want to move to UUIDs? Is that even possible with the number of
+        # places that rely on unit names being of the form {app_name}/{unit_number}?
+        # Note that CMRs don't expose the actual unit name, but still follow the form,
+        # using someething like remote-56a0f163eb4e4f2e88d50983dca7be02/0.
+        self.name = name
+        self.units = units
+
+    def __hash__(self):
+        # An Application instance from one relation will not be identical to one from another relation,
+        # since they may have different views of the units of that app. However, they should be considered
+        # equivalent, since they are views of the same entity. Generally, this should only matter for using
+        # model.app as an index into a given relation's relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Application):
+            raise TypeError(f"'==' not supported between instances of '{type(self).__name__}' and '{type(other).__name__}'")
+        return self.name == other.name
+
+
+class Unit:
+    def __init__(self, name):
+        # TODO Do we want to move to UUIDs? Is that even possible with the number of
+        # places that rely on unit names being of the form {app_name}/{unit_number}?
+        # Note that CMRs don't expose the actual unit name, but still follow the form,
+        # using someething like remote-56a0f163eb4e4f2e88d50983dca7be02/0.
+        self.name = name
+
+    def __hash__(self):
+        # A Unit instance from one relation will not be identical to one from another relation.
+        # However, they should be considered equivalent, since they are views of the same entity.
+        # Generally, this should only matter for using model.unit as an index into a given relation's
+        # relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Unit):
+            raise TypeError(f"'==' not supported between instances of '{type(self).__name__}' and '{type(other).__name__}'")
+        return self.name == other.name
+
+
+class RelationMap(dict):
+    """Map of relation names to lists of Relation instances.
+
+    The data for each relation name may be lazy-loaded.
+    """
+    def __init__(self, data):
+        super().__init__()
+        for relation_name, relations in data.items():
+            if relations is not None:
+                relations = [Relation(relation_id, relation_data)
+                             for relation_id, relation_data in sorted(relations.items())]
```

Comment:
I'm a bit confused about the behavior here. How would we get a list of relations upfront via the metadata?  Isn't that necessarily going to require a runtime introspection, currently via the command below?

Reply:
It is.  The pre-load support here is mainly in case we decide to go that route (e.g., if we get a "give me the world" hook tool from Juju) and to make testing easier.  And the tests I added were as much to give a feel for the interaction pattern as to verify the correctness.  If we want to drop tests and all the pre-load logic for now to simplify things, I'm happy to do so.

Reply:
Yeah, it'd be good to have something that feels more like the real path here. I'm happy to discuss how to make this in a way that is both fast and more realistic, and doesn't involve the weight of having to run external processes all the time during testing. Otherwise we're not only testing something that feels very little like the code we'll run for real, but we're also making the logic even more complex than usual. Double loss.